### PR TITLE
improve worm support

### PR DESCRIPTION
**Pull request by 27149chen**
# What problem are we solving?
https://github.com/seaweedfs/seaweedfs/issues/5954

# How are we solving the problem?
1. add a new field `Frozen` to filer `Entry`, it is a bool value, if true, the file is readonly and can not be deleted
2. if worm is enabled, set  `Frozen` to true when closing the file
3. when opening a file with write permission, if it is frezen, return EPERM
4. when deleting a file, if it is the last hard link, return EPERM
5. if worm is enabled in filer, write/delete an existing file will return "operation not permitted"

# How is the PR tested?

# Checks
- [ ] I have added unit tests if possible.
- [ ] I will add related wiki document changes and link to this PR after merging.

### CODE REVIEW

File: weed/mount/weedfs_file_mkrm.go [hunk 1]
```diff
@@ -110,6 +111,8 @@ func (wfs *WFS) Mknod(cancel <-chan struct{}, in *fuse.MknodIn, name string, out
 
 }
 
+var swapFileNameRegexp = regexp.MustCompile(`^\.([a-zA-Z0-9._-]+)\.swp$`)
```
Comment (chrislusf) [inline weed/mount/weedfs_file_mkrm.go]: Not generic.

**Review by chrislusf — CHANGES_REQUESTED** [review]

File: weed/mount/weedfs_file_mkrm.go (hunk 1 above)
Comment (27149chen) [reply weed/mount/weedfs_file_mkrm.go]: Do you have any suggestion?

File: weed/server/filer_server_handlers_write.go [hunk 2]
```diff
@@ -197,6 +196,10 @@ func (fs *FilerServer) move(ctx context.Context, w http.ResponseWriter, r *http.
 // curl -X DELETE http://localhost:8888/path/to?recursive=true&ignoreRecursiveError=true
 // curl -X DELETE http://localhost:8888/path/to?recursive=true&skipChunkDeletion=true
 func (fs *FilerServer) DeleteHandler(w http.ResponseWriter, r *http.Request) {
+	if fs.option.WORM {
```
Comment (half-life666) [inline weed/server/filer_server_handlers_write.go]: We still need a way to delete a WORM file. Consider a file is half written, there is no way to delete it if WORM is enabled.

There are 2 possible options:
1. set up a retention, after the retention, the file is deleted automatically, or can be deleted by DELETE API here. 
2. Implement a new cmd, or force option (or privilieged option, which can be used for special user), wich careful control

Without above, the only thing we can do is to delete entry from filemeta table?

File: weed/mount/weedfs_file_mkrm.go (hunk 1 above)
Comment (half-life666) [reply weed/mount/weedfs_file_mkrm.go]: Just document the consequences when worm is enabled? Then if we have a solution to delete the file (swap or whatever), provide that solution together in the document. 

If above are provided, we can delete the special hanling here?

File: weed/server/filer_server_handlers_write.go (hunk 2 above)
Comment (27149chen) [reply weed/server/filer_server_handlers_write.go]: add `?force=true`

**Comment by chrislusf** [comment]
Now I think a better design is to use https://github.com/seaweedfs/seaweedfs/wiki/Path-Specific-Configuration to configure the WORM. And It does not support specific files such as `.swp`.

**Comment by 27149chen** [comment]
> Now I think a better design is to use https://github.com/seaweedfs/seaweedfs/wiki/Path-Specific-Configuration to configure the WORM. And It does not support specific files such as `.swp`.

@chrislusf what do you mean by "It does not support specific files such as `.swp`", I can not create such files or they do not follow the rules?

File: weed/mount/weedfs_file_mkrm.go (hunk 1 above)
Comment (27149chen) [reply weed/mount/weedfs_file_mkrm.go]: removed. It seams that vi will create temp files not only .swp, but also .swx (x = [a-z]), and other names.
We should document that vi (and other editiors which will create temp files) is not allowed in WORM mode

**Comment by 27149chen** [comment]
> Now I think a better design is to use https://github.com/seaweedfs/seaweedfs/wiki/Path-Specific-Configuration to configure the WORM. And It does not support specific files such as `.swp`.

@chrislusf move the worm flag from filter cmd to fs configure, please take a look

File: weed/mount/filehandle_read.go [hunk 3]
```diff
@@ -47,6 +47,8 @@ func (fh *FileHandle) readFromChunks(buff []byte, offset int64) (int64, int64, e
 	if fileSize == 0 {
 		glog.V(1).Infof("empty fh %v", fileFullPath)
 		return 0, 0, io.EOF
+	} else if offset >= fileSize {
```
Comment (chrislusf) [inline weed/mount/filehandle_read.go]: unrelated change, separate the PR.

File: weed/filer/filechunk_group.go [hunk 4]
```diff
@@ -46,6 +47,9 @@ func (group *ChunkGroup) AddChunk(chunk *filer_pb.FileChunk) error {
 }
 
 func (group *ChunkGroup) ReadDataAt(fileSize int64, buff []byte, offset int64) (n int, tsNs int64, err error) {
+	if offset >= fileSize {
```
Comment (chrislusf) [inline weed/filer/filechunk_group.go]: unrelated change, separate the PR.

File: weed/server/filer_server_handlers_write.go [hunk 5]
```diff
@@ -197,7 +196,7 @@ func (fs *FilerServer) move(ctx context.Context, w http.ResponseWriter, r *http.
 // curl -X DELETE http://localhost:8888/path/to?recursive=true&ignoreRecursiveError=true
 // curl -X DELETE http://localhost:8888/path/to?recursive=true&skipChunkDeletion=true
 func (fs *FilerServer) DeleteHandler(w http.ResponseWriter, r *http.Request) {
-
+	force := r.URL.Query().Get("force") == "true"
```
Comment (chrislusf) [inline weed/server/filer_server_handlers_write.go]: this should not be needed since filer config can be changed

File: other/java/client/src/main/proto/filer.proto [hunk 6]
```diff
@@ -125,6 +125,7 @@ message Entry {
 
     RemoteEntry remote_entry = 10;
     int64 quota = 11; // for bucket only. Positive/Negative means enabled/disabled.
+    bool frozen = 12; // if true, the file is readonly and cannot be unfrozen.
```
Comment (chrislusf) [inline other/java/client/src/main/proto/filer.proto]: change to write_once_read_many here

File: other/java/client/src/main/proto/filer.proto [hunk 7]
```diff
@@ -442,6 +443,7 @@ message FilerConf {
         string data_node = 11;
         uint32 max_file_name_length = 12;
         bool disable_chunk_deletion = 13;
+        bool worm = 14;
```
Comment (chrislusf) [inline other/java/client/src/main/proto/filer.proto:445]: change to write_once_read_many here

File: weed/command/mount.go [hunk 8]
```diff
@@ -33,7 +33,7 @@ type MountOptions struct {
 	localSocket        *string
 	disableXAttr       *bool
 	extraOptions       []string
-	writeOnceReadMany  *bool
+	worm               *bool
```
Comment (chrislusf) [inline weed/command/mount.go]: change to writeOnceReadMany here

File: weed/command/mount.go [hunk 9]
```diff
@@ -71,7 +71,7 @@ func init() {
 	mountOptions.debugPort = cmdMount.Flag.Int("debug.port", 6061, "http port for debugging")
 	mountOptions.localSocket = cmdMount.Flag.String("localSocket", "", "default to /tmp/seaweedfs-mount-<mount_dir_hash>.sock")
 	mountOptions.disableXAttr = cmdMount.Flag.Bool("disableXAttr", false, "disable xattr")
-	mountOptions.writeOnceReadMany = cmdMount.Flag.Bool("writeOnceReadMany", false, "write once, read many times")
+	mountOptions.worm = cmdMount.Flag.Bool("worm", false, "wrom mode, if true, you can change a file only once, and then it is readonly and undeletable, see https://en.wikipedia.org/wiki/Write_once_read_many")
```
Comment (chrislusf) [inline weed/command/mount.go]: change to writeOnceReadMany here

File: weed/command/mount_std.go [hunk 10]
```diff
@@ -247,7 +248,7 @@ func RunMount(option *MountOptions, umask os.FileMode) bool {
 		Cipher:             cipher,
 		UidGidMapper:       uidGidMapper,
 		DisableXAttr:       *option.disableXAttr,
-		WriteOnceReadMany:  *option.writeOnceReadMany,
+		WORM:               *option.worm,
```
Comment (chrislusf) [inline weed/command/mount_std.go]: change to writeOnceReadMany here

File: weed/filer/entry.go [hunk 11]
```diff
@@ -43,6 +43,7 @@ type Entry struct {
 	Content         []byte
 	Remote          *filer_pb.RemoteEntry
 	Quota           int64
+	Frozen          bool
```
Comment (chrislusf) [inline weed/filer/entry.go]: change to WriteOnceReadMany here

File: weed/mount/weedfs_file_mkrm.go [hunk 12]
```diff
@@ -130,6 +130,11 @@ func (wfs *WFS) Unlink(cancel <-chan struct{}, header *fuse.InHeader, name strin
 		return code
 	}
 
+	// you can't delete a frozen file in worm mode
+	if wfs.option.WORM && entry != nil && entry.HardLinkCounter <= 1 && entry.Frozen {
```
Comment (chrislusf) [inline weed/mount/weedfs_file_mkrm.go]: need to prevent hard linking to a writeOnceReadMany file/dir

Comment (chrislusf) [reply weed/mount/weedfs_file_mkrm.go]: in here, wfs.option.WORM should not matter

File: weed/pb/filer.proto [hunk 13]
```diff
@@ -125,6 +125,7 @@ message Entry {
 
     RemoteEntry remote_entry = 10;
     int64 quota = 11; // for bucket only. Positive/Negative means enabled/disabled.
+    bool frozen = 12; // if true, the file is readonly and cannot be unfrozen.
```
Comment (chrislusf) [inline weed/pb/filer.proto]: change to write_once_read_many

File: weed/pb/filer.proto [hunk 14]
```diff
@@ -442,6 +443,7 @@ message FilerConf {
         string data_node = 11;
         uint32 max_file_name_length = 12;
         bool disable_chunk_deletion = 13;
+        bool worm = 14;
```
Comment (chrislusf) [inline weed/pb/filer.proto:445]: change to write_once_read_many

File: weed/mount/weedfs_file_mkrm.go (hunk 12 above)
Comment (27149chen) [reply weed/mount/weedfs_file_mkrm.go]: there is an example - a worm filesystem which support hardlink: https://docs.oracle.com/cd/E22586_01/html/E22571/gkxss.html

**Comment by 27149chen** [comment]
@chrislusf I think worm is better than writeOnceReadMany, because worm is a proper noun, it is a specific file system mode, but writeOnceReadMany is just a feature and cannot indicate that files like this cannot be deleted. see an example here https://docs.oracle.com/cd/E22586_01/html/E22571/gkxss.html

File: weed/command/mount.go (hunk 9 above)
Comment (chrislusf) [reply weed/command/mount.go]: fix spelling error "wrom mode"

Comment (chrislusf) [reply weed/command/mount.go]: "if true, you can change a file only once, and then it is readonly and undeletable"  simplify this.

File: weed/filer/entry.go [hunk 15]
```diff
@@ -43,6 +43,7 @@ type Entry struct {
 	Content         []byte
 	Remote          *filer_pb.RemoteEntry
 	Quota           int64
+	WORM            bool
```
Comment (chrislusf) [inline weed/filer/entry.go]: use "Worm"

**Comment by chrislusf** [comment]
Still I prefer "WriteOnceReadMany". No need to add one more layer of barrier to understand the code, even though it is simple or common.

**Comment by 27149chen** [comment]
But how do we know a file can not been deleted if we set "WriteOnceReadMany"?

File: weed/command/mount.go (hunk 9 above)
Comment (27149chen) [reply weed/command/mount.go]: > "if true, you can change a file only once, and then it is readonly and undeletable" simplify this.

I think it is simple enough

**Comment by chrislusf** [comment]
The design needs a big change.

* If using path specific config, the WriteOnceReadMany flag is not needed.
* However, this would not protect against hardlinks crossing the path boundary. So the hardlinks crossing the path boundary should not be allowed.

**Comment by 27149chen** [comment]
ok, I'll remove hardlink support, we can implement it later
But why "If using path specific config, the WriteOnceReadMany flag is not needed."?

**Comment by chrislusf** [comment]
> But why "If using path specific config, the WriteOnceReadMany flag is not needed."?

Only use the path-specific config to determine WriteOnceReadMany flag. When the path-specific config removes the flag, the files would become normal again and no need to "force" delete them.

**Comment by 27149chen** [comment]
> > But why "If using path specific config, the WriteOnceReadMany flag is not needed."?
> 
> Only use the path-specific config to determine WriteOnceReadMany flag. When the path-specific config removes the flag, the files would become normal again and no need to "force" delete them.

You means the worm flag in weed mount command?

**Comment by chrislusf** [comment]
> You mean the worm flag in weed mount command?

No. https://github.com/seaweedfs/seaweedfs/wiki/Path-Specific-Configuration

**Comment by 27149chen** [comment]
> > You mean the worm flag in weed mount command?
> 
> No. https://github.com/seaweedfs/seaweedfs/wiki/Path-Specific-Configuration

the worm flag is not needed in path-specific config? I cannot understand, can you explain how to implement worm features using current path-specific options?

**Comment by chrislusf** [comment]
the worm flag is not needed in file entry.

**Comment by 27149chen** [comment]
> the worm flag is not needed in file entry.

if worm flag is removed from file entry, how do I know if write is allowed when opening a file?

**Comment by chrislusf** [comment]
> if worm flag is removed from file entry, how do I know if write is allowed when opening a file?

check the path-specific configuration

**Comment by 27149chen** [comment]
current flow:
1. first open: writable
2. first close(flush): set worm flag in file entry
3. second open: readonly (by checking the worm flag in file entry)

If worm flag is removed ,how do I know if it is the first open? Can you describe this process?

**Comment by chrislusf** [comment]
If it is first open, it would be a new entry and there would not be existing entry.

**Comment by 27149chen** [comment]
> If it is first open, it would be a new entry and there would not be existing entry.

entry is created in Mknod，Open is to open an existing entry

**Comment by 27149chen** [comment]
> Check created time equals modified time.

Is the accuracy of these two times in seconds not precise enough?

**Comment by chrislusf** [comment]
Should be enough. Also check 0 file size.

**Comment by 27149chen** [comment]
I think a dedicated flag is better than guessing the state by other attributes, a write may finish in one second, also zero size is not equal to the first write

**Comment by 27149chen** [comment]
echo "1" > /xxx
result:
```
"Mtime": "2024-09-12T12:44:23+08:00",
"Crtime": "2024-09-12T12:44:23+08:00",
```

**Comment by chrislusf** [comment]
A flag would need more special handling when need to change back to normal file.

**Comment by chrislusf** [comment]
> echo "1" > /xxx
> 
> result:
> 
> ```
> 
> "Mtime": "2024-09-12T12:44:23+08:00",
> 
> "Crtime": "2024-09-12T12:44:23+08:00",
> 
> ```

File size is not zero.

**Comment by 27149chen** [comment]
Is it possible to parallelly write a new file? In such case, opening a new file might cause non-zero file size

**Comment by chrislusf** [comment]
> Is it possible to parallelly write a new file? In such case, opening a new file might cause non-zero file size

The purpose for WriteOnceReadMany is to protect against future modifications. This edge case, assuming it is possible, is out of scope.

**Comment by 27149chen** [comment]
changes:
1. remove worm flag in Entry, if fileSize != 0 and crtime != mtime, it is not a new file
2. load latest filer conf when weed mount starts, and subscribe new changes
3. check if a file is in a worm enbaled path by rules in filer conf

File: weed/server/filer_server_handlers_write_autochunk.go [hunk 16]
```diff
@@ -121,6 +126,10 @@ func (fs *FilerServer) doPutAutoChunk(ctx context.Context, w http.ResponseWriter
 		contentType = ""
 	}
 
+	if err := fs.checkPermissions(ctx, r, ""); err != nil {
```
Comment (jerebear12) [inline weed/server/filer_server_handlers_write_autochunk.go:129]: Is passing this empty string intentional?

Comment (27149chen) [reply weed/server/filer_server_handlers_write_autochunk.go:129]: I think it is a mistake, pr: https://github.com/seaweedfs/seaweedfs/pull/6142
